lib/Doctrine/ORM/Proxy/ClassUtils.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Proxy;
+
+use Doctrine\Persistence\Proxy;
+
+use function get_class;
+use function strrpos;
+use function substr;
+
+/**
+ * Class-related functionality for objects that might or not be proxy objects
+ * at the moment.
+ *
+ * @internal
```

Comment:
Should this be an internal class ? The need to present the actual class name in error message or similar things also exists outside the ORM.
The class in doctrine/common was not internal.

Maybe this class should be moved to doctrine/persistence near the proxy interface.

Reply:
> Maybe this class should be moved to doctrine/persistence near the proxy interface.

It might be best, here are the usages: https://github.com/search?q=org%3Adoctrine%20ClassUtils&type=code

- `doctrine/orm`, obviously
- `DoctrineBundle` also uses it
- `orientdb-odm` as well, whatever that is
- `phpcr-odm`

The mongoDB ODM seems to have moved away from it, they say so in their UPGRADE file. Seeing that reminded me of something @alcaeus did in `doctrine/persistence`: he contributed code very similar to what I'm contributing here, maybe we can reuse that: https://github.com/doctrine/persistence/pull/145

Reply:
Looking at doctrine/persistence#145 a bit more, it seems to me that the way to go would to have the `ProxyClassNameResolver` implementation live in the ORM. That way, `doctrine/persistence` would not have a public default implementation. I think it has a private one purely for BC reasons. Once we have that implementation in the ORM, we could inject it into the inside `ClassMetadataFactory::__construct()`, ensuring we use the same implementation in the code of `doctrine/persistence` as well, without relying on copy/paste.



> The drawback of the class name resolver is that it requires access to the Configuration class.

Looking at the usages inside `doctrine/orm`, the Configuration instance is available in most places, so I don't think that would be a big deal.

Reply:
@stof I pushed a new implementation that relies on what I think you had in mind. Let me know if that would suit you and I will port something similar to 2.17.x (since there is no breaking change).